From eb1a3cafdd49ec72fc53da1141c64c73713631f2 Mon Sep 17 00:00:00 2001
From: Takashi Yano <takashi.yano@nifty.ne.jp>
Date: Mon, 18 Apr 2022 20:43:44 +0900
Subject: [PATCH 061/N] Cygwin: pty: Fix Ctrl-C behaviour in latest GDB.

- In the latest GDB (11.2-1), Ctrl-C behaviour is broken a bit for
  non-cygwin inferior. For example, Ctrl-C on GDB prompt is not sent
  to GDB but to the inferior. This patch fixes the issue.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 winsup/cygwin/fhandler_termios.cc | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/winsup/cygwin/fhandler_termios.cc b/winsup/cygwin/fhandler_termios.cc
index daec184..0d53ad2 100644
--- a/winsup/cygwin/fhandler_termios.cc
+++ b/winsup/cygwin/fhandler_termios.cc
@@ -349,7 +349,8 @@ fhandler_termios::process_sigs (char c, tty* ttyp, fhandler_termios *fh)
 			 cygwin apps started from non-cygwin shell. */
       if (c == '\003' && p && p->ctty == ttyp->ntty && p->pgid == pgid
 	  && ((p->process_state & PID_NOTCYGWIN)
-	      || (p->process_state & PID_NEW_PG)
+	      || ((p->process_state & PID_NEW_PG)
+		  && ttyp->pcon_input_state_eq (tty::to_nat))
 	      || !(p->process_state & PID_CYGPARENT)))
 	{
 	  /* Ctrl-C event will be sent only to the processes attaching
@@ -423,6 +424,7 @@ fhandler_termios::process_sigs (char c, tty* ttyp, fhandler_termios *fh)
 	    with_debugger = true; /* inferior is cygwin app */
 	  if (!(p->process_state & PID_NOTCYGWIN)
 	      && (p->process_state & PID_NEW_PG) /* Check marker */
+	      && ttyp->pcon_input_state_eq (tty::to_nat)
 	      && p->pid == pgid)
 	    with_debugger_nat = true; /* inferior is non-cygwin app */
 	}
-- 
2.9.0

